data/generate_embeddings.py

```diff
@@ -0,0 +1,598 @@
+DEFAULT_RETRY_MIN_WAIT = 1
+DEFAULT_RETRY_MAX_WAIT = 30
+
+
+
+# -------------------------
+# Text cleaning & tokenization
+# -------------------------
+def basic_clean(text: str) -> str:
+    """Clean and normalize text content."""
+    if not text:
+        return ""
+    # Normalize whitespace
+    text = re.sub(r"[\r\t]", " ", text)
+    text = re.sub(r"\u00A0", " ", text)
+    # Collapse multiple spaces/newlines
+    text = re.sub(r"\s+", " ", text).strip()
+    # Remove non-informative symbols (keep punctuation, letters, numbers)
+    text = re.sub(r"[^\w\s\.,;:!\?\-\(\)/]", "", text)
+    return text
+
+
+def get_encoder():
+    """Get tiktoken encoder if available."""
+    if not TIKTOKEN_AVAILABLE:
+        return None
+    try:
+        return tiktoken.get_encoding("cl100k_base")
+    except Exception as e:
+        print(f"Error initializing tiktoken encoder: {e}")
+        return None
+
+
+def tokenize_text(text: str, encoder) -> List[int]:
+    """Tokenize text using encoder or fallback to word-based."""
+    if encoder is None:
+        # Fallback: split by words
+        return text.split()
+    return encoder.encode(text)
+
+
+def detokenize_tokens(tokens: List[Any], encoder) -> str:
+    """Convert tokens back to text."""
+    if encoder is None:
+        if isinstance(tokens, list):
+            return " ".join(tokens)
+        return str(tokens)
+    return encoder.decode(tokens)
+
+
+def sliding_window_chunk_tokens(
+    tokens: List[Any], 
+    max_tokens: int, 
+    overlap_ratio: float
+) -> List[List[Any]]:
+    """Create overlapping chunks using sliding window approach."""
+    if max_tokens <= 0:
+        return []
+    if not tokens:
+        return []
+    
+    overlap = int(max_tokens * overlap_ratio)
+    overlap = min(overlap, max_tokens - 1) if max_tokens > 1 else 0
+
+    chunks: List[List[Any]] = []
+    start = 0
+    
+    while start < len(tokens):
+        end = min(start + max_tokens, len(tokens))
+        chunks.append(tokens[start:end])
+        
+        if end == len(tokens):
+            break
+            
+        start = end - overlap if overlap > 0 else end
+    
+    return chunks
+
+# -------------------------
+# I/O utils
+# -------------------------
+def read_file_content(path: Path) -> str:
+    """Read and parse file content based on file type."""
+    suffix = path.suffix.lower()
+    
+    try:
+        text = path.read_text(encoding="utf-8")
+    except UnicodeDecodeError:
+        text = path.read_text(encoding="latin-1", errors="ignore")
+    
+    if suffix == ".json":
+        try:
+            data = json.loads(text)
+            # If JSON has a top-level "text" field, prefer that
+            if isinstance(data, dict) and "text" in data and isinstance(data["text"], str):
+                return data["text"]
+            # Otherwise, stringify
+            return json.dumps(data, ensure_ascii=False)
+        except Exception:
+            return text
+    
+    return text
```

Comment:
When reading JSON files, catch `UnicodeDecodeError` inside the JSON block for better error reporting.
Log file read errors with more context.